### PR TITLE
Update the VS AnalyzerDependencyChecker

### LINKED ISSUE #2684: Fix VS IDE live analysis to replicate analyzer behavior with no explicit dependencies
Git commit #2502 changed VBCSCompiler.exe to reject builds where an analyzer dependency is not explicitly specified. The fallback for this is to use the standalone csc.exe/vbc.exe which runs in a separate process and uses `Assembly.LoadFrom` to load analyzer assemblies. So, if the analyzer dependency is actually packaged with the analyzer and present in the same folder, the fallback path successfully loads and executes the analyzer.

However, our `VisualStudioAnalyzer` type for executing analyzers in IDE live analysis fails for this scenario as it uses the `ShadowCopyAnalyzerAssemblyLoader` : http://source.roslyn.io/#Microsoft.VisualStudio.LanguageServices/Implementation/ProjectSystem/MetadataReferences/VisualStudioAnalyzerAssemblyLoaderService.cs,ee9d36e9927f4be7.

This causes a bad IDE experience where the live analysis generates analyzer exception diagnostics such as below:

```
Severity	Code	Description	Project	File	Line
Warning	CS8032	An instance of analyzer Analyzer5.Analyzer5Analyzer cannot be created from C:\Users\mavasani\Documents\Visual Studio 2015\Projects\Analyzer5\Analyzer5\Analyzer5\bin\Debug\Analyzer5.dll: Exception has been thrown by the target of an invocation..	ClassLibrary1		1
```

While the command line build succeeds, reporting actual analyzer diagnostics, and the error list ends up with both analyzer load failure diagnostics from live analysis + analyzer reported diagnostics from build:

analyzer_ide_inconsistency

@tmeschter is aware about this issue, and plans to fix it soon. This issue tracks this work.

### LINKED ISSUE #2728: Unhelpful warning when an analyzer could not be loaded because another version of analyzer was loaded for a different project
1. Create a C# diagnostic + code fix project (I have "Analyzer5" for this repro) and build solution.

2. In a new VS instance, create another C# diagnostic + code fix project (I have "Analyzer6" for this repro).

3. Rename the output assembly name and default namespace for Analyzer6 project to "Analyzer5" through project property pages. Build solution.

4. In a new VS instance, create a new solution with 2 C# projects, Classlibrary1.csproj and Classlibrary2.csproj

5. Add nuget package "Analyzer5.1.0.0.0.nupkg" to Classlibrary1.csproj to add the analyzer created in step (1).

6. Add nuget package "Analyzer6.1.0.0.0.nupkg" to Classlibrary2.csproj to add the analyzer created in step (3).

**Actual**: Following diagnostics show in the error list.

```
Severity	Code	Description	Project	File	Line
Warning	Analyzer5	Type name 'Class1' contains lowercase letters	ClassLibrary1	C:\Users\mavasani\Documents\Visual Studio 2015\Projects\ClassLibrary2\ClassLibrary1\Class1.cs	8

Warning	CS8032	An instance of analyzer Analyzer5.Analyzer6Analyzer cannot be created from C:\Users\mavasani\Documents\Visual Studio 2015\Projects\ClassLibrary2\packages\Analyzer6.1.0.0.0\tools\analyzers\Analyzer5.dll: Value cannot be null.
Parameter name: type.	ClassLibrary2		1

```

**Expected:**
Given that both the nuget packages have an analyzer with the same name, but different content, it is expected that only one copy of Analyzer5.dll gets loaded in VS and hence analyzer for only one of the projects runs fine. However, the CS8032 generated for Classlibrary2 is not too helpful as it does not indicate anything about the core issue: _"projects have analyzer references to assemblies with same assembly identity but different contents"_

Note that an informative and helpful `AnalyzerDependencyConflict` diagnostic **does** indeed show up if the same mismatch occurs for an analyzer dependency:

```
Severity	Code	Description	Project	File	Line
Warning	AnalyzerDependencyConflict	Assembly 'C:\Users\mavasani\Documents\Visual Studio 2015\Projects\ClassLibrary2\packages\Analyzer5.1.0.0.0\tools\analyzers\AnalyzerDependency_2.dll' used by analyzer 'AnalyzerDependency_1' and assembly 'C:\Users\mavasani\Documents\Visual Studio 2015\Projects\ClassLibrary2\packages\Analyzer6.1.0.0.0\tools\analyzers\AnalyzerDependency_2.dll' used by analyzer 'AnalyzerDependency_1' have the same identity but different contents. These analyzers may not run correctly.	ClassLibrary1		1
```

We should probably have such a diagnostic for the core analyzer assembly conflict as well.

### LINKED ISSUE #2738: Duplicate AnalyzerDependencyConflict diagnostics and possibly unnecessary analyzer assembly loading just to detect dependency conflicts
1. Create a C# diagnostic + code fix project (I have "Analyzer5" for this repro).

2. Add 2 portable class library projects to the solution, say "AnalyzerDependency_1" and "AnalyzerDependency_2". Add P2P reference from AnalyzerDependency_1 to AnalyzerDependency_2. Add the below code snippets to respective projects:

    1. AnalyzerDependency_2:
        ```c#
namespace AnalyzerDependency_2
{
    public class Class1
    {
        public string Message = "Message";
    }
}
        ```

    2. AnalyzerDependency_1:
        ```c#
namespace AnalyzerDependency_1
{
    public class Class1
    {
        public string Title = "Title";
        public static string Message = new AnalyzerDependency_2.Class1().Message;
    }
}
        ```

3. Add P2P reference from primary Analyzer5 project to AnalyzerDependency_1.

4. Build solution.

5. Repeat all the steps 1-4 above on a new solution (say Analyzer6). However, use different strings for code added to AnalyzerDependency_* projects, say "Title2" and "Message2". This ensures that Analyzer5 and Analyzer6 have dependencies with same assembly name, but different contents.

6. In a new VS instance, create a new solution with 2 C# projects, Classlibrary1.csproj and Classlibrary2.csproj

7. Add nuget package "Analyzer5.1.0.0.0.nupkg" to Classlibrary1.csproj to add the analyzer created in step (1).

8. Add nuget package "Analyzer6.1.0.0.0.nupkg" to Classlibrary2.csproj to add the analyzer created in step (5).

**Actual**: Following duplicate `AnalyzerDependencyConflict` diagnostics show in the error list.

```
Severity	Code	Description	Project	File	Line
Warning	AnalyzerDependencyConflict	Assembly 'C:\Users\mavasani\Documents\Visual Studio 2015\Projects\ClassLibrary2\packages\Analyzer5.1.0.0.0\tools\analyzers\AnalyzerDependency_2.dll' used by analyzer 'AnalyzerDependency_1' and assembly 'C:\Users\mavasani\Documents\Visual Studio 2015\Projects\ClassLibrary2\packages\Analyzer6.1.0.0.0\tools\analyzers\AnalyzerDependency_2.dll' used by analyzer 'AnalyzerDependency_1' have the same identity but different contents. These analyzers may not run correctly.	ClassLibrary1		1

Warning	AnalyzerDependencyConflict	Assembly 'C:\Users\mavasani\Documents\Visual Studio 2015\Projects\ClassLibrary2\packages\Analyzer5.1.0.0.0\tools\analyzers\AnalyzerDependency_2.dll' used by analyzer 'AnalyzerDependency_1' and assembly 'C:\Users\mavasani\Documents\Visual Studio 2015\Projects\ClassLibrary2\packages\Analyzer6.1.0.0.0\tools\analyzers\AnalyzerDependency_2.dll' used by analyzer 'AnalyzerDependency_1' have the same identity but different contents. These analyzers may not run correctly.	ClassLibrary2		1
```

It would be good to de-duplicate such diagnostics across projects.

Probably the more severe issue here is: are we actually loading AnalyzerDependency_1.dll and AnalyzerDependency_2.dll? Neither of these 2 dependency assemblies have any DiagnosticAnalyzers, so shouldn't be loaded to instantiate analyzers from it. Additionally, neither of these 2 assemblies are actually referenced by Analyzer5.dll and Analyzer6.dll (they ended up in the nuget package only because they were dumped to the output folder of core analyzer projects due to a redundant P2P reference).

So, are we eagerly loading analyzer reference assemblies just to determine if they have dependency conflicts, when the assembly itself has no analyzers and hence shouldn't be ever loaded?

### PR DESCRIPTION
**Pull request by tmeschter**
This commit updates the `AnalyzerDependencyChecker` type to reflect the
recent changes to require all analyzers and dependencies to be
specified, and the changes in how we load analyzer assemblies.

Now that we require all analyzers and dependencies to be specified, we
no longer look next to the specified analyzers to optimistically locate
dependencies. Instead, we only consider the list of passed-in analyzers
*plus* the assemblies already loaded into the `AppDomain`. The latter
avoids unhelpful warnings in the event that the analyzer assembly
includes a code fix that take a dependency on a Visual Studio assembly,
which of course can't be included in the list of analyzers. We now
create a warning when we identify a missing dependency.

We now also check analyzers themselves for conflicts (i.e., assemblies
at different paths with the same identity but different contents)
instead of just their dependencies.

Fixes #2684.
Fixes #2728.
Fixes #2738.

@mavasani @srivatsn @heejaechang @jmarolf @shyamnamboodiripad @ManishJayaswal Could you take a look, please?

### CODE REVIEW

**Comment by jmarolf** [comment]
:+1:

**Comment by tmeschter** [comment]
@mavasani I'm going to close the three issues you opened with this change, so if you could take a look, I would appreciate that.

File: src/VisualStudio/Core/Def/ServicesVSResources.resx [hunk 1]
```diff
@@ -480,4 +480,10 @@ Use the dropdown to view and switch to other projects this file may belong to.</
   <data name="WorkspaceOutputPaneTitle" xml:space="preserve">
     <value>IntelliSense</value>
   </data>
+  <data name="WRN_MissingAnalyzerReferenceId" xml:space="preserve">
+    <value>MissingAnalyzerReference</value>
```
Comment (mavasani) [inline src/VisualStudio/Core/Def/ServicesVSResources.resx]: Do we need to localize the ID? Probably yes given its a description of the symptom.

File: src/VisualStudio/Core/Def/ServicesVSResources.resx [hunk 2]
```diff
@@ -480,4 +480,10 @@ Use the dropdown to view and switch to other projects this file may belong to.</
...
+  </data>
+  <data name="WRN_MissingAnalyzerReferenceMessage" xml:space="preserve">
+    <value>Analyzer assembly '{0}' depends on '{1}' but it was not found. Analyzers may not run correctly.</value>
```
Comment (mavasani) [inline src/VisualStudio/Core/Def/ServicesVSResources.resx]: Thanks, this should be helpful!

**Comment by mavasani** [comment]
:+1: Thanks for fixing the error messages too!

**Comment by heejaechang** [comment]
:+1:

**Comment by tmeschter** [comment]
@dotnet-bot Retest this, please.

**Pull request merged**